src/components/views/rooms/DecryptionFailureBar.tsx

```diff
@@ -0,0 +1,247 @@
+    const [waiting, setWaiting] = useState<boolean>(true);
+    useEffect(() => {
+        const timeout = setTimeout(() => setWaiting(false), WAIT_PERIOD);
+        return () => clearTimeout(timeout);
+    }, []);
+
+    // Is this device unverified?
+    const [needsVerification, setNeedsVerification] = useState<boolean>(false);
+    // Does this user have verified devices other than this device?
+    const [hasOtherVerifiedDevices, setHasOtherVerifiedDevices] = useState<boolean>(false);
+    // Does this user have key backups?
+    const [hasKeyBackup, setHasKeyBackup] = useState<boolean>(false);
+
+    // Keep track of session IDs that do not have pending key requests
+    const [unrequestedSessions, setUnrequestedSessions] = useState<Set<string>>(new Set());
+
+    useEffect(() => {
+        let cancelled = false;
+        const cancel = () => {
+            cancelled = true;
+        };
+        const updateSessions = async () => {
+            // Aggregate session IDs of undecryptable messages, and check
+            // which ones do not have pending key requests
+            const currentRequestedSessions = new Set<string>();
+            const currentUnrequestedSessions = new Set<string>();
+            for (const event of failures) {
+                const wireContent = event.getWireContent();
+                const sessionId = wireContent.session_id;
+                if (!sessionId
+                    || currentRequestedSessions.has(sessionId)
+                    || currentUnrequestedSessions.has(sessionId)
+                ) continue;
+                if (await context.getOutgoingRoomKeyRequest(event)) {
+                    currentRequestedSessions.add(sessionId);
+                } else {
+                    currentUnrequestedSessions.add(sessionId);
+                }
+                if (cancelled) return;
+            }
+            setUnrequestedSessions(currentUnrequestedSessions);
+        };
+        updateSessions();
+        return cancel;
+    }, [context, failures, setUnrequestedSessions]);
+
+    const resendKeyRequests = useCallback(async () => {
+        // Resend key requests for any sessions that do not currently have pending requests
+        const oldUnrequestedSessions = new Set(unrequestedSessions);
+        setUnrequestedSessions(new Set());
+
+        const newRequestedSessions = new Set<string>();
+        for (const event of failures) {
+            const sessionId = event.getWireContent().session_id;
+            if (!sessionId || !oldUnrequestedSessions.has(sessionId) || newRequestedSessions.has(sessionId)) continue;
+            newRequestedSessions.add(sessionId);
+            await context.cancelAndResendEventRoomKeyRequest(event);
+        }
+    }, [context, unrequestedSessions, failures]);
+
+    // Recheck which devices are verified and whether we have key backups
+    const updateDeviceInfo = useCallback(async () => {
+        await context.downloadKeys([context.getUserId()], true);
```

Comment:
We're not doing this every time it gets updated, only when it first gets rendered. From a perspective of avoiding user frustration, I think it's important that when we're giving a call-to-action in order to resolve an error, it should be based on accurate information, and that downloading keys once when this component is first rendered is worthwhile in that regard.

Reply:
> only when it first gets rendered

well still, every time the user changes room.

I'm really, really dubious about this. If there are bugs in the device list cache invalidation, we should fix them, rather than papering over them by just ignoring the cache (which will do nothing to fix other bugs that result from incorrect caches, and make it harder to reason about the system). We've been down this path with room-key-requests: they helped paper over some of the bugs, but the underlying problems still bite and room key requests make them harder to track down and fix properly.

Reply:
> I'm not saying that there are bugs in cache invalidation. I'm saying that the state of our other devices and keys may have changed since the last time the cache was invalidated, 

That would be a bug in the cache invalidation, though?

Reply:
We may be operating under different mental models here? I'm imagining
- Device A refreshes its cache of the user's devices/keys
- Device B signs out
- Device A doesn't know that Device B has signed out, nobody's proactively telling Device A that this has happened, so Device A won't have that updated information until its cache expires
- Device A displays this decryption failure bar before that cache expires, and gives the user incorrect directions because it assumes Device B is still logged in

No part of that is a bug in cache invalidation. Unless you think that Device B and/or the homeserver should send a message to Device A whenever any device logs in or out or changes verification state? Which is definitely outside the scope of this PR.

Reply:
> Device A doesn't know that Device B has signed out, nobody's proactively telling Device A that this has happened, so Device A won't have that updated information until its cache expires

The server should proactively tell device A that Device B has signed out, via the `device_lists` field in `/sync`.

Reply:
Fixed; you're right, this is much cleaner :)